// @base-ui-components/react/menu/store/MenuStore.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MenuStore = void 0;
var React = _interopRequireWildcard(require("react"));
var _store = require("@base-ui-components/utils/store");
var _empty = require("@base-ui-components/utils/empty");
var _useRefWithInit = require("@base-ui-components/utils/useRefWithInit");
var _getEmptyRootContext = require("../../floating-ui-react/utils/getEmptyRootContext");
var _FloatingTreeStore = require("../../floating-ui-react/components/FloatingTreeStore");
const selectors = {
  open: (0, _store.createSelector)(state => state.open),
  disabled: (0, _store.createSelector)(state => state.parent.type === 'menubar' ? state.parent.context.disabled || state.disabled : state.disabled),
  modal: (0, _store.createSelector)(state => (state.parent.type === undefined || state.parent.type === 'context-menu') && (state.modal ?? true)),
  mounted: (0, _store.createSelector)(state => state.mounted),
  activeTriggerId: (0, _store.createSelector)(state => state.activeTriggerId),
  activeTriggerElement: (0, _store.createSelector)(state => state.mounted && state.activeTriggerId != null ? state.triggers.get(state.activeTriggerId) ?? null : null),
  isTriggerActive: (0, _store.createSelector)((state, triggerId) => triggerId !== undefined && state.activeTriggerId === triggerId),
  isOpenedByTrigger: (0, _store.createSelector)((state, triggerId) => triggerId !== undefined && state.activeTriggerId === triggerId && state.open),
  allowMouseEnter: (0, _store.createSelector)(state => state.parent.type === 'menu' ? state.parent.store.select('allowMouseEnter') : state.allowMouseEnter),
  stickIfOpen: (0, _store.createSelector)(state => state.stickIfOpen),
  parent: (0, _store.createSelector)(state => state.parent),
  rootId: (0, _store.createSelector)(state => {
    if (state.parent.type === 'menu') {
      return state.parent.store.select('rootId');
    }
    return state.parent.type !== undefined ? state.parent.context.rootId : state.rootId;
  }),
  activeIndex: (0, _store.createSelector)(state => state.activeIndex),
  isActive: (0, _store.createSelector)((state, itemIndex) => state.activeIndex === itemIndex),
  hoverEnabled: (0, _store.createSelector)(state => state.hoverEnabled),
  positionerElement: (0, _store.createSelector)(state => state.positionerElement),
  transitionStatus: (0, _store.createSelector)(state => state.transitionStatus),
  instantType: (0, _store.createSelector)(state => state.instantType),
  lastOpenChangeReason: (0, _store.createSelector)(state => state.lastOpenChangeReason),
  floatingRootContext: (0, _store.createSelector)(state => state.floatingRootContext),
  floatingTreeRoot: (0, _store.createSelector)(state => {
    if (state.parent.type === 'menu') {
      return state.parent.store.select('floatingTreeRoot');
    }
    return state.floatingTreeRoot;
  }),
  floatingNodeId: (0, _store.createSelector)(state => state.floatingNodeId),
  floatingParentNodeId: (0, _store.createSelector)(state => state.floatingParentNodeId),
  itemProps: (0, _store.createSelector)(state => state.itemProps),
  popupProps: (0, _store.createSelector)(state => state.popupProps),
  activeTriggerProps: (0, _store.createSelector)(state => state.activeTriggerProps),
  inactiveTriggerProps: (0, _store.createSelector)(state => state.inactiveTriggerProps),
  payload: (0, _store.createSelector)(state => state.payload),
  triggers: (0, _store.createSelector)(state => state.triggers),
  closeDelay: (0, _store.createSelector)(state => state.closeDelay),
  keyboardEventRelay: (0, _store.createSelector)(state => {
    if (state.keyboardEventRelay) {
      return state.keyboardEventRelay;
    }
    if (state.parent.type === 'menu') {
      return state.parent.store.select('keyboardEventRelay');
    }
    return undefined;
  })
};
class MenuStore extends _store.ReactStore {
  constructor(initialState) {
    super({
      ...createInitialState(),
      ...initialState
    }, {
      positionerRef: /*#__PURE__*/React.createRef(),
      popupRef: /*#__PURE__*/React.createRef(),
      typingRef: {
        current: false
      },
      itemDomElements: {
        current: []
      },
      itemLabels: {
        current: []
      },
      allowMouseUpTriggerRef: {
        current: false
      },
      preventUnmountingRef: {
        current: false
      },
      triggerFocusTargetRef: /*#__PURE__*/React.createRef(),
      beforeContentFocusGuardRef: /*#__PURE__*/React.createRef(),
      onOpenChangeComplete: undefined
    }, selectors);

    // Sync `allowMouseEnter` with parent menu if applicable.
    this.observe((0, _store.createSelector)(state => state.allowMouseEnter), (allowMouseEnter, oldValue) => {
      // The allowMouseEnter !== oldValue check prevent calling parent store's set
      // on intialization. Without it, React might complain about updating one component during rendering another.
      if (this.state.parent.type === 'menu' && allowMouseEnter !== oldValue) {
        this.state.parent.store.set('allowMouseEnter', allowMouseEnter);
      }
    });

    // Set up propagation of state from parent menu if applicable.
    this.unsubscribeParentListener = this.observe('parent', parent => {
      this.unsubscribeParentListener?.();
      if (parent.type === 'menu') {
        this.unsubscribeParentListener = parent.store.subscribe(() => {
          this.notifyAll();
        });
        this.context.allowMouseUpTriggerRef = parent.store.context.allowMouseUpTriggerRef;
        return;
      }
      if (parent.type !== undefined) {
        this.context.allowMouseUpTriggerRef = parent.context.allowMouseUpTriggerRef;
      }
      this.unsubscribeParentListener = null;
    });
  }
  setOpen(open, eventDetails) {
    this.state.floatingRootContext.events.emit('setOpen', {
      open,
      eventDetails
    });
  }
  static useStore(externalStore, initialState) {
    // eslint-disable-next-line react-hooks/rules-of-hooks
    const store = (0, _useRefWithInit.useRefWithInit)(() => {
      return externalStore ?? new MenuStore(initialState);
    }).current;
    return store;
  }
  unsubscribeParentListener = null;
}
exports.MenuStore = MenuStore;
function createInitialState() {
  return {
    open: false,
    disabled: false,
    modal: true,
    mounted: false,
    allowMouseEnter: true,
    stickIfOpen: true,
    parent: {
      type: undefined
    },
    rootId: undefined,
    activeIndex: null,
    hoverEnabled: true,
    positionerElement: null,
    transitionStatus: 'idle',
    instantType: undefined,
    lastOpenChangeReason: null,
    floatingRootContext: (0, _getEmptyRootContext.getEmptyRootContext)(),
    floatingTreeRoot: new _FloatingTreeStore.FloatingTreeStore(),
    floatingNodeId: undefined,
    floatingParentNodeId: null,
    itemProps: _empty.EMPTY_OBJECT,
    popupProps: _empty.EMPTY_OBJECT,
    activeTriggerProps: _empty.EMPTY_OBJECT,
    inactiveTriggerProps: _empty.EMPTY_OBJECT,
    payload: undefined,
    triggers: new Map(),
    activeTriggerId: null,
    keyboardEventRelay: undefined,
    closeDelay: 0
  };
}
