// @base-ui-components/react/esm/popover/store/PopoverStore.js
/* eslint-disable react-hooks/rules-of-hooks */
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { ReactStore, createSelector } from '@base-ui-components/utils/store';
import { Timeout } from '@base-ui-components/utils/useTimeout';
import { useRefWithInit } from '@base-ui-components/utils/useRefWithInit';
import { useOnMount } from '@base-ui-components/utils/useOnMount';
import { EMPTY_OBJECT } from '@base-ui-components/utils/empty';
import { getEmptyRootContext } from "../../floating-ui-react/utils/getEmptyRootContext.js";
import { REASONS } from "../../utils/reasons.js";
import { PATIENT_CLICK_THRESHOLD } from "../../utils/constants.js";
function createInitialState() {
  return {
    open: false,
    mounted: false,
    modal: false,
    activeTriggerId: null,
    positionerElement: null,
    popupElement: null,
    triggers: new Map(),
    instantType: undefined,
    transitionStatus: 'idle',
    openMethod: null,
    openReason: null,
    titleElementId: undefined,
    descriptionElementId: undefined,
    floatingRootContext: getEmptyRootContext(),
    payload: undefined,
    activeTriggerProps: EMPTY_OBJECT,
    inactiveTriggerProps: EMPTY_OBJECT,
    popupProps: EMPTY_OBJECT,
    stickIfOpen: true,
    nested: false
  };
}
const selectors = {
  open: createSelector(state => state.open),
  mounted: createSelector(state => state.mounted),
  activeTriggerId: createSelector(state => state.activeTriggerId),
  activeTriggerElement: createSelector(state => state.mounted && state.activeTriggerId != null ? state.triggers.get(state.activeTriggerId) ?? null : null),
  positionerElement: createSelector(state => state.positionerElement),
  popupElement: createSelector(state => state.popupElement),
  triggers: createSelector(state => state.triggers),
  instantType: createSelector(state => state.instantType),
  transitionStatus: createSelector(state => state.transitionStatus),
  openMethod: createSelector(state => state.openMethod),
  openReason: createSelector(state => state.openReason),
  modal: createSelector(state => state.modal),
  stickIfOpen: createSelector(state => state.stickIfOpen),
  floatingRootContext: createSelector(state => state.floatingRootContext),
  titleElementId: createSelector(state => state.titleElementId),
  descriptionElementId: createSelector(state => state.descriptionElementId),
  payload: createSelector(state => state.payload),
  activeTriggerProps: createSelector(state => state.activeTriggerProps),
  inactiveTriggerProps: createSelector(state => state.inactiveTriggerProps),
  popupProps: createSelector(state => state.popupProps)
};
export class PopoverStore extends ReactStore {
  constructor(initialState) {
    super({
      ...createInitialState(),
      ...initialState
    }, {
      popupRef: /*#__PURE__*/React.createRef(),
      backdropRef: /*#__PURE__*/React.createRef(),
      internalBackdropRef: /*#__PURE__*/React.createRef(),
      onOpenChange: undefined,
      onOpenChangeComplete: undefined,
      triggerFocusTargetRef: /*#__PURE__*/React.createRef(),
      beforeContentFocusGuardRef: /*#__PURE__*/React.createRef(),
      preventUnmountingRef: {
        current: false
      },
      stickIfOpenTimeout: new Timeout()
    }, selectors);
  }
  setOpen = (nextOpen, eventDetails) => {
    const isHover = eventDetails.reason === REASONS.triggerHover;
    const isKeyboardClick = eventDetails.reason === REASONS.triggerPress && eventDetails.event.detail === 0;
    const isDismissClose = !nextOpen && (eventDetails.reason === REASONS.escapeKey || eventDetails.reason == null);
    eventDetails.preventUnmountOnClose = () => {
      this.context.preventUnmountingRef.current = true;
    };
    this.context.onOpenChange?.(nextOpen, eventDetails);
    if (eventDetails.isCanceled) {
      return;
    }
    const details = {
      open: nextOpen,
      nativeEvent: eventDetails.event,
      reason: eventDetails.reason,
      nested: this.state.nested,
      triggerElement: eventDetails.trigger
    };
    const floatingEvents = this.state.floatingRootContext.events;
    floatingEvents?.emit('openchange', details);
    const changeState = () => {
      this.set('open', nextOpen);
      if (nextOpen) {
        this.set('openReason', eventDetails.reason ?? null);
      }

      // If a popup is closing, the `trigger` may be null.
      // We want to keep the previous value so that exit animations are played and focus is returned correctly.
      const newTriggerId = eventDetails.trigger?.id ?? null;
      if (newTriggerId || nextOpen) {
        this.set('activeTriggerId', newTriggerId);
      }
    };
    if (isHover) {
      // Only allow "patient" clicks to close the popover if it's open.
      // If they clicked within 500ms of the popover opening, keep it open.
      this.set('stickIfOpen', true);
      this.context.stickIfOpenTimeout.start(PATIENT_CLICK_THRESHOLD, () => {
        this.set('stickIfOpen', false);
      });
      ReactDOM.flushSync(changeState);
    } else {
      changeState();
    }
    if (isKeyboardClick || isDismissClose) {
      this.set('instantType', isKeyboardClick ? 'click' : 'dismiss');
    } else if (eventDetails.reason === REASONS.focusOut) {
      this.set('instantType', 'focus');
    } else {
      this.set('instantType', undefined);
    }
  };
  static useStore(externalStore, initialState) {
    const store = useRefWithInit(() => {
      return externalStore ?? new PopoverStore(initialState);
    }).current;
    useOnMount(store.disposeEffect);
    return store;
  }
  disposeEffect = () => {
    return this.context.stickIfOpenTimeout.disposeEffect();
  };
}
