// @base-ui-components/react/esm/tooltip/store/TooltipStore.js
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { createSelector, ReactStore } from '@base-ui-components/utils/store';
import { EMPTY_OBJECT } from '@base-ui-components/utils/empty';
import { useRefWithInit } from '@base-ui-components/utils/useRefWithInit';
import { getEmptyRootContext } from "../../floating-ui-react/utils/getEmptyRootContext.js";
import { REASONS } from "../../utils/reasons.js";
const selectors = {
  open: createSelector(state => state.open),
  mounted: createSelector(state => state.mounted),
  disabled: createSelector(state => state.disabled),
  instantType: createSelector(state => state.instantType),
  isInstantPhase: createSelector(state => state.isInstantPhase),
  floatingRootContext: createSelector(state => state.floatingRootContext),
  trackCursorAxis: createSelector(state => state.trackCursorAxis),
  transitionStatus: createSelector(state => state.transitionStatus),
  disableHoverablePopup: createSelector(state => state.disableHoverablePopup),
  preventUnmountingOnClose: createSelector(state => state.preventUnmountingOnClose),
  lastOpenChangeReason: createSelector(state => state.lastOpenChangeReason),
  triggers: createSelector(state => state.triggers),
  activeTriggerId: createSelector(state => state.activeTriggerId),
  activeTriggerElement: createSelector(state => state.mounted && state.activeTriggerId != null ? state.triggers.get(state.activeTriggerId) ?? null : null),
  activeTriggerProps: createSelector(state => state.activeTriggerProps),
  inactiveTriggerProps: createSelector(state => state.activeTriggerProps),
  payload: createSelector(state => state.payload),
  popupProps: createSelector(state => state.popupProps),
  popupElement: createSelector(state => state.popupElement),
  positionerElement: createSelector(state => state.positionerElement)
};
export class TooltipStore extends ReactStore {
  constructor(initialState) {
    super({
      ...createInitialState(),
      ...initialState
    }, {
      popupRef: /*#__PURE__*/React.createRef(),
      onOpenChange: undefined,
      onOpenChangeComplete: undefined
    }, selectors);
  }
  setOpen = (nextOpen, eventDetails) => {
    const reason = eventDetails.reason;
    const isHover = reason === REASONS.triggerHover;
    const isFocusOpen = nextOpen && reason === REASONS.triggerFocus;
    const isDismissClose = !nextOpen && (reason === REASONS.triggerPress || reason === REASONS.escapeKey);
    eventDetails.preventUnmountOnClose = () => {
      this.set('preventUnmountingOnClose', true);
    };
    this.context.onOpenChange?.(nextOpen, eventDetails);
    if (eventDetails.isCanceled) {
      return;
    }
    const changeState = () => {
      if (isFocusOpen) {
        this.set('instantType', 'focus');
      } else if (isDismissClose) {
        this.set('instantType', 'dismiss');
      } else if (reason === REASONS.triggerHover) {
        this.set('instantType', undefined);
      }
      this.update({
        open: nextOpen,
        lastOpenChangeReason: reason
      });
      const newTriggerId = eventDetails.trigger?.id ?? null;
      if (newTriggerId || nextOpen) {
        this.set('activeTriggerId', newTriggerId);
      }
    };
    if (isHover) {
      // If a hover reason is provided, we need to flush the state synchronously. This ensures
      // `node.getAnimations()` knows about the new state.
      ReactDOM.flushSync(changeState);
    } else {
      changeState();
    }
  };
  static useStore(externalStore, initialState) {
    // eslint-disable-next-line react-hooks/rules-of-hooks
    return useRefWithInit(() => {
      return externalStore ?? new TooltipStore(initialState);
    }).current;
  }
}
function createInitialState() {
  return {
    open: false,
    mounted: false,
    disabled: false,
    instantType: undefined,
    isInstantPhase: true,
    floatingRootContext: getEmptyRootContext(),
    trackCursorAxis: 'none',
    transitionStatus: 'idle',
    disableHoverablePopup: false,
    preventUnmountingOnClose: false,
    lastOpenChangeReason: null,
    triggers: new Map(),
    payload: undefined,
    activeTriggerId: null,
    activeTriggerProps: EMPTY_OBJECT,
    inactiveTriggerProps: EMPTY_OBJECT,
    popupProps: EMPTY_OBJECT,
    popupElement: null,
    positionerElement: null
  };
}
