// @base-ui-components/react/esm/dialog/store/DialogStore.js
import * as React from 'react';
import { createSelector, ReactStore } from '@base-ui-components/utils/store';
import { EMPTY_OBJECT } from '@base-ui-components/utils/empty';
import { getEmptyRootContext } from "../../floating-ui-react/utils/getEmptyRootContext.js";
const selectors = {
  open: createSelector(state => state.open),
  modal: createSelector(state => state.modal),
  nested: createSelector(state => state.nested),
  nestedOpenDialogCount: createSelector(state => state.nestedOpenDialogCount),
  disablePointerDismissal: createSelector(state => state.disablePointerDismissal),
  openMethod: createSelector(state => state.openMethod),
  descriptionElementId: createSelector(state => state.descriptionElementId),
  titleElementId: createSelector(state => state.titleElementId),
  mounted: createSelector(state => state.mounted),
  transitionStatus: createSelector(state => state.transitionStatus),
  popupProps: createSelector(state => state.popupProps),
  floatingRootContext: createSelector(state => state.floatingRootContext),
  activeTriggerId: createSelector(state => state.activeTriggerId),
  activeTriggerElement: createSelector(state => state.mounted && state.activeTriggerId != null ? state.triggers.get(state.activeTriggerId) ?? null : null),
  triggers: createSelector(state => state.triggers),
  popupElement: createSelector(state => state.popupElement),
  viewportElement: createSelector(state => state.viewportElement),
  payload: createSelector(state => state.payload),
  activeTriggerProps: createSelector(state => state.activeTriggerProps),
  inactiveTriggerProps: createSelector(state => state.inactiveTriggerProps),
  role: createSelector(state => state.role)
};
export class DialogStore extends ReactStore {
  constructor(initialState) {
    super(createInitialState(initialState), {
      popupRef: /*#__PURE__*/React.createRef(),
      backdropRef: /*#__PURE__*/React.createRef(),
      internalBackdropRef: /*#__PURE__*/React.createRef(),
      preventUnmountingOnCloseRef: {
        current: false
      }
    }, selectors);
  }
  setOpen = (nextOpen, eventDetails) => {
    eventDetails.preventUnmountOnClose = () => {
      this.context.preventUnmountingOnCloseRef.current = true;
    };
    if (!nextOpen && eventDetails.trigger == null && this.state.activeTriggerId != null) {
      // When closing the dialog, pass the old trigger to the onOpenChange event
      // so it's not reset too early (potentially causing focus issues in controlled scenarios).
      eventDetails.trigger = this.state.triggers.get(this.state.activeTriggerId);
    }
    this.context.onOpenChange?.(nextOpen, eventDetails);
    if (eventDetails.isCanceled) {
      return;
    }
    const details = {
      open: nextOpen,
      nativeEvent: eventDetails.event,
      reason: eventDetails.reason,
      nested: this.state.nested
    };
    this.state.floatingRootContext.events?.emit('openchange', details);
    this.set('open', nextOpen);
    const newTriggerId = eventDetails.trigger?.id ?? null;
    if (newTriggerId || nextOpen) {
      this.set('activeTriggerId', newTriggerId);
    }
  };
}
function createInitialState(initialState = {}) {
  return {
    disablePointerDismissal: false,
    modal: true,
    open: false,
    nested: false,
    popupElement: null,
    viewportElement: null,
    activeTriggerId: null,
    descriptionElementId: undefined,
    titleElementId: undefined,
    openMethod: null,
    mounted: false,
    transitionStatus: 'idle',
    nestedOpenDialogCount: 0,
    triggers: new Map(),
    floatingRootContext: getEmptyRootContext(),
    payload: undefined,
    activeTriggerProps: EMPTY_OBJECT,
    inactiveTriggerProps: EMPTY_OBJECT,
    popupProps: EMPTY_OBJECT,
    role: 'dialog',
    ...initialState
  };
}
