// @base-ui-components/react/esm/popover/root/PopoverRoot.js
'use client';

import * as React from 'react';
import { useStableCallback } from '@base-ui-components/utils/useStableCallback';
import { useIsoLayoutEffect } from '@base-ui-components/utils/useIsoLayoutEffect';
import { useScrollLock } from '@base-ui-components/utils/useScrollLock';
import { useDismiss, useFloatingRootContext, useInteractions, useRole, FloatingTree, useFloatingParentNodeId } from "../../floating-ui-react/index.js";
import { useTransitionStatus } from "../../utils/useTransitionStatus.js";
import { useOpenChangeComplete } from "../../utils/useOpenChangeComplete.js";
import { PopoverRootContext, usePopoverRootContext } from "./PopoverRootContext.js";
import { PopoverStore } from "../store/PopoverStore.js";
import { createChangeEventDetails } from "../../utils/createBaseUIEventDetails.js";
import { REASONS } from "../../utils/reasons.js";
import { jsx as _jsx } from "react/jsx-runtime";
function PopoverRootComponent({
  props
}) {
  const {
    children,
    open: openProp,
    defaultOpen: defaultOpenProp = false,
    onOpenChange,
    onOpenChangeComplete,
    modal = false,
    handle,
    triggerId: triggerIdProp,
    defaultTriggerId: defaultTriggerIdProp = null
  } = props;
  const store = PopoverStore.useStore(handle?.store, {
    open: openProp ?? defaultOpenProp,
    modal,
    activeTriggerId: triggerIdProp !== undefined ? triggerIdProp : defaultTriggerIdProp
  });
  store.useControlledProp('open', openProp, defaultOpenProp);
  store.useControlledProp('activeTriggerId', triggerIdProp, defaultTriggerIdProp);
  const open = store.useState('open');
  const triggerElements = store.useState('triggers');
  const activeTriggerId = store.useState('activeTriggerId');
  const positionerElement = store.useState('positionerElement');
  const activeTriggerElement = store.useState('activeTriggerElement');
  const payload = store.useState('payload');
  const openReason = store.useState('openReason');
  const openMethod = store.useState('openMethod');
  const {
    mounted,
    setMounted,
    transitionStatus
  } = useTransitionStatus(open);
  let resolvedTriggerId = null;
  if (mounted === true && triggerIdProp === undefined && triggerElements.size === 1) {
    resolvedTriggerId = triggerElements.keys().next().value || null;
  } else {
    resolvedTriggerId = activeTriggerId ?? null;
  }
  store.useContextCallback('onOpenChange', onOpenChange);
  store.useContextCallback('onOpenChangeComplete', onOpenChangeComplete);
  useIsoLayoutEffect(() => {
    store.set('mounted', mounted);
    if (!mounted) {
      store.set('activeTriggerId', null);
    }
  }, [store, mounted]);
  useIsoLayoutEffect(() => {
    if (open) {
      store.set('activeTriggerId', resolvedTriggerId);
      if (resolvedTriggerId == null) {
        store.set('payload', undefined);
      }
    }
  }, [store, resolvedTriggerId, open]);
  useScrollLock(open && modal === true && openReason !== REASONS.triggerHover && openMethod !== 'touch', positionerElement);
  React.useEffect(() => {
    if (!open) {
      store.context.stickIfOpenTimeout.clear();
    }
  }, [store, open]);
  const createPopoverEventDetails = React.useCallback(reason => {
    const details = createChangeEventDetails(reason);
    details.preventUnmountOnClose = () => {
      store.context.preventUnmountingRef.current = true;
    };
    return details;
  }, [store]);
  const handleUnmount = useStableCallback(() => {
    setMounted(false);
    store.update({
      stickIfOpen: true,
      openReason: null,
      activeTriggerId: null,
      mounted: false
    });
    onOpenChangeComplete?.(false);
  });
  const handleImperativeClose = React.useCallback(() => {
    store.setOpen(false, createPopoverEventDetails(REASONS.imperativeAction));
  }, [store, createPopoverEventDetails]);
  useOpenChangeComplete({
    enabled: !store.context.preventUnmountingRef.current,
    open,
    ref: store.context.popupRef,
    onComplete() {
      if (!open) {
        handleUnmount();
      }
    }
  });
  React.useImperativeHandle(props.actionsRef, () => ({
    unmount: handleUnmount,
    close: handleImperativeClose
  }), [handleUnmount, handleImperativeClose]);
  const floatingContext = useFloatingRootContext({
    elements: {
      reference: activeTriggerElement,
      floating: positionerElement,
      triggers: Array.from(triggerElements.values())
    },
    open,
    onOpenChange: store.setOpen
  });
  const dismiss = useDismiss(floatingContext, {
    outsidePressEvent: {
      // Ensure `aria-hidden` on outside elements is removed immediately
      // on outside press when trapping focus.
      mouse: modal === 'trap-focus' ? 'sloppy' : 'intentional',
      touch: 'sloppy'
    }
  });
  const role = useRole(floatingContext);
  const {
    getReferenceProps,
    getFloatingProps,
    getTriggerProps
  } = useInteractions([dismiss, role]);
  store.useSyncedValues({
    transitionStatus,
    modal,
    activeTriggerProps: getReferenceProps(),
    inactiveTriggerProps: getTriggerProps(),
    popupProps: getFloatingProps(),
    floatingRootContext: floatingContext,
    nested: useFloatingParentNodeId() != null
  });
  const popoverContext = React.useMemo(() => ({
    store
  }), [store]);
  return /*#__PURE__*/_jsx(PopoverRootContext.Provider, {
    value: popoverContext,
    children: typeof children === 'function' ? children({
      payload
    }) : children
  });
}

/**
 * Groups all parts of the popover.
 * Doesn’t render its own HTML element.
 *
 * Documentation: [Base UI Popover](https://base-ui.com/react/components/popover)
 */
export function PopoverRoot(props) {
  if (usePopoverRootContext(true)) {
    return /*#__PURE__*/_jsx(PopoverRootComponent, {
      props: props
    });
  }
  return /*#__PURE__*/_jsx(FloatingTree, {
    children: /*#__PURE__*/_jsx(PopoverRootComponent, {
      props: props
    })
  });
}
